### PR TITLE
UD flush 

### PR DESCRIPTION
**Pull request by alex-mikheev**
@yosefe please take a look

### CODE REVIEW

File: src/uct/ib/ud/base/ud_ep.c [hunk 1]
```diff
@@ -453,15 +459,40 @@ void uct_ud_ep_process_rx(uct_ud_iface_t *iface, uct_ud_neth_t *neth, unsigned b
     ucs_mpool_put(skb);
 }
 
+ucs_status_t uct_ud_ep_flush_do(uct_ud_iface_t *iface, uct_ud_ep_t *ep)
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_ep.c]: maybe  rename tpo `uct_ud_ep_flush_nolock()`

File: src/uct/ib/ud/base/ud_ep.c [hunk 2]
```diff
@@ -108,6 +108,12 @@ static UCS_CLASS_CLEANUP_FUNC(uct_ud_ep_t)
 
     ucs_arbiter_group_purge(&iface->tx.pending_q, &self->tx.pending.group,
                             uct_ud_ep_pending_cancel_cb, 0);
+
+    if (ucs_queue_is_empty(&self->tx.window)) {
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_ep.c]: should be !empty?

File: src/uct/ib/ud/base/ud_ep.c [hunk 3]
```diff
@@ -453,15 +459,40 @@ void uct_ud_ep_process_rx(uct_ud_iface_t *iface, uct_ud_neth_t *neth, unsigned b
...
+    if (skb->neth->packet_type & UCT_UD_PACKET_FLAG_ACK_REQ) {
+        /* last packet was already sent with ack request. 
+         * either by flush or 
+         * Do not send more, let reqular retransmission 
+         * mechanism do the work
+         */
+        return UCS_INPROGRESS;
+    }
+
+    skb->neth->packet_type |= UCT_UD_PACKET_FLAG_ACK_REQ;
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_ep.c]: changing flag for an in-flight packet might have unpredictable behavior

Comment (yosefe) [inline src/uct/ib/ud/base/ud_ep.c]: also, probably should add valgrind macros to post_send/send_completion to catch this

File: src/uct/ib/ud/base/ud_ep.h [hunk 4]
```diff
@@ -225,27 +227,20 @@ void uct_ud_ep_process_rx(uct_ud_iface_t *iface,
...
 }
 
 static UCS_F_ALWAYS_INLINE void
-uct_ud_neth_ctl_ack(uct_ud_ep_t *ep, uct_ud_neth_t *neth)
+uct_ud_neth_ctl_ack_req(uct_ud_ep_t *ep, uct_ud_neth_t *neth)
 {
     neth->psn         = ep->tx.psn;
     neth->ack_psn     = ep->rx.acked_psn = ucs_frag_list_sn(&ep->rx.ooo_pkts);
-    neth->packet_type = ep->dest_ep_id;
+    neth->packet_type = (ep->dest_ep_id|UCT_UD_PACKET_FLAG_ACK_REQ);
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_ep.h]: no need for ( ) here

File: src/uct/ib/ud/base/ud_ep.h [hunk 5]
```diff
@@ -313,6 +302,31 @@ static UCS_F_ALWAYS_INLINE int uct_ud_ep_no_window(uct_ud_ep_t *ep)
...
+ */
+static UCS_F_ALWAYS_INLINE int uct_ud_ep_req_ack(uct_ud_ep_t *ep)
+{
+    uct_ud_psn_t acked_psn, max_psn, psn;
+
+    max_psn   = ep->tx.max_psn;
+    acked_psn = ep->tx.acked_psn;
+    psn       = ep->tx.psn;
+
+    return UCT_UD_PSN_COMPARE(psn, ==, ((acked_psn*3 + max_psn)>>2)) ||
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_ep.h]: please add spaces around operators

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1143/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/561/ for details (Mellanox internal link).

**Comment by shamisp** [comment]
autobot build

**Comment by shamisp** [comment]
```
 CXX    uct/gtest-test_amo.o
cc1plus: warnings being treated as errors
/ucx-build/gtest-1.7.0/include/gtest/gtest.h: In function 'testing::AssertionResult testing::internal::CmpHelperEQ(const char*, const char*, const T1&, const T2&) [with T1 = int, T2 = long unsigned int]':
/ucx-build/gtest-1.7.0/include/gtest/gtest.h:1525:   instantiated from 'static testing::AssertionResult testing::internal::EqHelper<true>::Compare(const char*, const char*, const T1&, const T2&, typename testing::internal::EnableIf<(! testing::internal::is_pointer<T2>::value)>::type*) [with T1 = int, T2 = size_t]'
uct/test_ud.cc:46:   instantiated from here
/ucx-build/gtest-1.7.0/include/gtest/gtest.h:1448: warning: comparison between signed and unsigned integer expressions
make[2]: *** [uct/gtest-test_ud.o] Error 1
```

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1146/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/563/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
:+1:

**Pull request merged**